rust/src/smb/detect.rs

```diff
@@ -201,3 +204,258 @@ pub extern "C" fn rs_smb_tx_get_dce_iface(state: &mut SMBState,
+        std::mem::drop(Box::from_raw(ptr as *mut SmbCmdData));
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct SmbCmdData (HashMap<u8, HashSet<u16>>);
+
+impl SmbCmdData {
+
+    fn new(cmd_codes1: HashSet<u16>, cmd_codes2: HashSet<u16>) -> Self {
+
+        let mut cmd_data = HashMap::new();
+        cmd_data.insert(1, cmd_codes1);
+        cmd_data.insert(2, cmd_codes2);
+
+        return Self(cmd_data);
+    }
+
+}
+
+fn str_to_u16(v: &str) -> Result<u16, ()> {
+    let size;
+    if v.starts_with("0x") {
+        let no_prefix = v.trim_start_matches("0x");
+        size = u16::from_str_radix(&no_prefix, 16);
+    } else {
+        size = u16::from_str_radix(&v, 10);
+    }
+
+    return size.map_err(|_| ());
+}
+
+
+fn parse_cmd_data(arg: &str) -> Result<SmbCmdData, ()> {
+
+    let cmd_names1 = gen_smb1_command_names();
+    let cmd_names2 = gen_smb2_command_names();
+
+
+    let split_args: Vec<&str> = arg.split(',').collect();
+
+    let mut cmd_codes1 = HashSet::new();
+    let mut cmd_codes2 = HashSet::new();
+
+    for cmd in split_args.iter() {
+        let cmd = cmd.trim();
+
+        match str_to_u16(cmd) {
+            Ok(cmd_code) =>  {
+                cmd_codes1.insert(cmd_code);
+                cmd_codes2.insert(cmd_code);
+            }
+            Err(_) => {
+                let mut in_any = false;
+                if let Some(cmd_code) = cmd_names1.get(cmd) {
+                    cmd_codes1.insert(*cmd_code);
+                    in_any = true;
+                }
+
+                if let Some(cmd_code) = cmd_names2.get(cmd) {
+                    cmd_codes2.insert(*cmd_code);
+                    in_any = true;
+                }
+
+                if !in_any {
+                    return Err(());
+                }
+            }
+        }
+
+    }
+    return Ok(SmbCmdData::new(cmd_codes1, cmd_codes2));
+}
+
+fn gen_smb2_command_names() -> HashMap<&'static str, u16> {
+    let mut cmd_names2 = HashMap::new();
+    cmd_names2.insert("negotiate", 0u16);
+    cmd_names2.insert("session_setup", 1u16);
+    cmd_names2.insert("logoff", 2u16);
+    cmd_names2.insert("tree_connect", 3u16);
+    cmd_names2.insert("tree_disconnect", 4u16);
+    cmd_names2.insert("create", 5u16);
+    cmd_names2.insert("close", 6u16);
+    cmd_names2.insert("flush", 7u16);
+    cmd_names2.insert("read", 8u16);
+    cmd_names2.insert("write", 9u16);
+    cmd_names2.insert("lock", 0xau16);
+    cmd_names2.insert("ioctl", 0xbu16);
+    cmd_names2.insert("cancel", 0xcu16);
+    cmd_names2.insert("echo", 0xdu16);
+    cmd_names2.insert("query_directory", 0xeu16);
+    cmd_names2.insert("change_notify", 0xfu16);
+    cmd_names2.insert("query_info", 0x10u16);
+    cmd_names2.insert("set_info", 0x11u16);
+    cmd_names2.insert("oplock_break", 0x12u16);
+
+    return cmd_names2;
+}
+
+fn gen_smb1_command_names() -> HashMap<&'static str, u16> {
+    let mut cmd_names1 = HashMap::new();
+    cmd_names1.insert("create_directory", 0x00u16);
```

Comment:
These do not look like the logged ones like `SMB1_COMMAND_NEGOTIATE_PROTOCOL` or `SMB1_COMMAND_SESSION_SETUP_ANDX`

There is some code to be reused somewhere I guess